clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/services/SubnetSelector.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2018 Lookout, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.ecs.services;
+
+import com.netflix.spinnaker.clouddriver.aws.cache.Keys;
+import com.netflix.spinnaker.clouddriver.aws.model.AmazonSubnet;
+import com.netflix.spinnaker.clouddriver.aws.provider.view.AmazonSubnetProvider;
+import com.netflix.spinnaker.clouddriver.ecs.model.EcsSubnet;
+import com.netflix.spinnaker.clouddriver.ecs.provider.view.AmazonPrimitiveConverter;
+import com.netflix.spinnaker.clouddriver.ecs.provider.view.EcsAccountMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Component
+public class SubnetSelector {
+
+  AmazonSubnetProvider amazonSubnetProvider;
+  AmazonPrimitiveConverter converter;
+  EcsAccountMapper ecsAccountMapper;
+
+  @Autowired
+  public SubnetSelector(AmazonSubnetProvider amazonSubnetProvider,
+                        AmazonPrimitiveConverter converter,
+                        EcsAccountMapper ecsAccountMapper) {
+    this.amazonSubnetProvider = amazonSubnetProvider;
+    this.converter = converter;
+    this.ecsAccountMapper = ecsAccountMapper;
+  }
+
+  public Collection<String> resolveSubnetsIds(String ecsAccountName, String region, String subnetType) {
+    String correspondingAwsAccountName = ecsAccountMapper.fromEcsAccountNameToAws(ecsAccountName).getName();
+
+    Set<AmazonSubnet> amazonSubnets = amazonSubnetProvider.getAllMatchingKeyPattern(
+      Keys.getSubnetKey("*", region, correspondingAwsAccountName));
+
+    Set<EcsSubnet> ecsSubnets = converter.convertToEcsSubnet(amazonSubnets);
+
+      Set<EcsSubnet> filteredSubnets = ecsSubnets
+      .stream()
+      .filter(subnet -> subnetType.equals(subnet.getPurpose()))
+      .collect(Collectors.toSet());
+
+    Set<String> filteredSubnetIds = filteredSubnets.stream().map(AmazonSubnet::getId).collect(Collectors.toSet());
```

Comment:
How about this?

```
ecsSubnets
  .stream()
  .filter(subnet -> stuff)
  .map(AmazonSubnet::getId)
  .collect(Collectors.toSet());
```

It doesn't look to me like you need to shuffle the stream into a set, then map the elements and collect it into yet another new set. If you absolutely need to get unique elements in between, it's better to `distinct()` on the stream.